File: combinatorics/resource/ST/project_5/bai_toan_4/general_graph/am_to_al.cpp
#include <iostream>
#include <vector>

using namespace std;

// Ham chuyen AM sang AL
vector<vector<int>> amToAl(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<int>> al(n);

    for (int u = 0; u < n; ++u) {
        for (int v = 0; v < n; ++v) {
            if (matrix[u][v] == 1) {
                al[u].push_back(v); // Neu co canh tu u den v thi them v vao danh sach ke cua u
            }
        }
    }

    return al;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));

    cout << "Nhap ma tran ke (AM):\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> matrix[i][j];

    vector<vector<int>> al = amToAl(matrix);

    cout << "Danh sach ke (AL):\n";
    for (int u = 0; u < n; ++u) {
        cout << u << ": ";
        for (int v : al[u])
            cout << v << " ";
        cout << "\n";
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/general_graph/am_to_al.py
def am_to_al(matrix):
    n = len(matrix)
    al = [[] for _ in range(n)]  # Khoi tao danh sach ke

    for u in range(n):
        for v in range(n):
            if matrix[u][v] == 1:  # Neu co canh tu u den v
                al[u].append(v)   # Them v vao danh sach cua u

    return al

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
print("Nhap ma tran ke (AM):")
am = []
for _ in range(n):
    row = list(map(int, input().split()))
    am.append(row)

al = am_to_al(am)

print("Danh sach ke (AL):")
for u in range(n):
    print(f"{u}: {' '.join(map(str, al[u]))}")